File: r/2015/day08a/day08a.r
#!/usr/bin/env Rscript

usage <- function() {
    args <- commandArgs()
    progname <- sub("^--file=", "", args[grep("^--file=", args)])
    cat("usage:", progname, "<input file>\n")
    quit(status = 1)
}

process <- function(filename) {
    result <- 0
    lines <- suppressWarnings(readLines(filename))
    for ( line in lines ) {
        codeLen <- nchar(line)
        memLen <- 0
        i <- 2
        while ( i <= nchar(line) ) {
            if ( substr(line, i, i) == '\\' ) {
                if ( substr(line, i + 1, i + 1) %in% c('\\', '"') ) {
                    i <- i + 2
                } else if ( substr(line, i + 1, i + 1) == 'x' ) {
                    i <- i + 4
                } else {
                    i <- i + 1
                }
            } else {
                i <- i + 1
            }
            memLen <- memLen + 1
        }
        result <- result + codeLen - memLen
    }
    return(result)
}

main <- function() {
    args <- commandArgs(trailingOnly = TRUE)
    if ( length(args) < 1 ) {
        usage()
    }
    filename <- args[1]
    result <- process(filename)
    cat("result =", result, "\n")
}

main()

File: dart/2015/day08a/day08a.dart
#!/usr/bin/env dart

import 'dart:io';

void usage(String progName) {
    print('usage: $progName <input file>');
    exit(1);
}

int process(String filename) {
    int result = 0;

    final lines = File(filename).readAsLinesSync();
    for ( final line in lines ) {
        final codeLen = line.length;
        int memLen = 0;
        int i = 1;
        while ( i < line.length - 1 ) {
            final ch1 = line.substring(i, i + 1);
            switch ( ch1 ) {
            case '\\':
                final ch2 = line.substring(i + 1, i + 2);
                switch( ch2 ) {
                case '\\':
                case '\"':
                    i += 2;
                    break;
                case 'x':
                    i += 4;
                    break;
                default:
                    i += 1;
                }
                break;
            default:
                i += 1;
            }
            memLen += 1;
        }
        result += codeLen - memLen;
    }
    return result;
}

void main(List<String> args) {
    if ( args.isEmpty ) {
        usage(Platform.script.pathSegments.last);
    }
    final filename = args[0];
    final result = process(filename);
    print('result = $result');
}
